React/Modules/RCTDevSettings.mm

```diff
@@ -0,0 +1,572 @@
+{
+  return [self initWithDefaultValues:nil];
+}
+
+- (instancetype)initWithDefaultValues:(NSDictionary *)defaultValues
+{
+  if (self = [super init]) {
+    if (defaultValues) {
+      [self _reloadWithDefaults:defaultValues];
+    }
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)updateSettingWithValue:(id)value forKey:(NSString *)key
+{
+  RCTAssert((key != nil), @"%@", [NSString stringWithFormat:@"%@: Tried to update nil key", [self class]]);
+
+  id currentValue = _settings[key];
+  if (currentValue == value || [currentValue isEqual:value]) {
+    return;
+  }
+  if (value) {
+    _settings[key] = value;
+  } else {
+    [_settings removeObjectForKey:key];
+  }
+  if (!_keysChanged) {
+    _keysChanged = [NSMutableSet set];
+  }
+  [_keysChanged addObject:key];
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+- (id)settingForKey:(NSString *)key
+{
+  return _settings[key];
+}
+
+- (void)_reloadWithDefaults:(NSDictionary *)defaultValues
+{
+  NSDictionary *existingSettings = [[NSUserDefaults standardUserDefaults] objectForKey:kRCTDevSettingsUserDefaultsKey];
+  _settings = (existingSettings) ? [existingSettings mutableCopy] : [NSMutableDictionary dictionary];
+  for (NSString *key in [defaultValues keyEnumerator]) {
+    if (!_settings[key]) {
+      _settings[key] = defaultValues[key];
+    }
+  }
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+@end
+
+@interface RCTDevSettings () <RCTBridgeModule, RCTInvalidating, RCTWebSocketObserverDelegate>
+{
+  NSURLSessionDataTask *_liveReloadUpdateTask;
+  NSURL *_liveReloadURL;
+  BOOL _isJSLoaded;
+}
+
+@property (nonatomic, strong) Class executorClass;
+@property (nonatomic, readwrite, strong) id<RCTDevSettingsDataSource> dataSource;
+
+@end
+
+@implementation RCTDevSettings
+
+@synthesize bridge = _bridge;
+
+RCT_EXPORT_MODULE()
+
+- (instancetype)init
+{
+  // default behavior is to use NSUserDefaults
+  NSDictionary *defaultValues = @{
+    kRCTDevSettingShakeToShowDevMenu: @YES,
+  };
+  RCTDevSettingsUserDefaultsDataSource *dataSource = [[RCTDevSettingsUserDefaultsDataSource alloc] initWithDefaultValues:defaultValues];
+  return [self initWithDataSource:dataSource];
+}
+
+- (instancetype)initWithDataSource:(id<RCTDevSettingsDataSource>)dataSource
+{
+  if (self = [super init]) {
+    _dataSource = dataSource;
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(jsLoaded:)
+                                                 name:RCTJavaScriptDidLoadNotification
+                                               object:nil];
+
+    // Delay setup until after Bridge init
+    __weak RCTDevSettings *weakSelf = self;
```

Comment:
No need for weakify here, it's fine for us to keep the object alive during a `dispatch_async(dispatch_get_main_queue()`